File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int size=nums.length;
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        for(int i=0;i<size;++i){
            if(i!=0 && nums[i]==nums[i-1]) continue;
            int j=i+1;
            int k=size-1;
            int target = 0-nums[i];
            while(j<k){
                if(nums[j]+nums[k]==target){
                    int[] triplet = new int[]{nums[i], nums[j], nums[k]};
                    res.add(Arrays.stream(triplet).boxed().toList());
                    while(j<k && nums[j]==triplet[1]){
                       ++j;
                   }
                   while(j<k && nums[k]==triplet[2]){
                     --k;
                    }
                }
                else if((nums[j]+nums[k])<target){
                    ++j;
                }
                else{
                    --k;
                }
            }
        }
        return res;
    }
}

File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n=nums.size();
        if(n==0) return {};
        vector<vector<int>>res;
        // [1,0,1,2,-1,-4]
        // [-4,-1,0,1,1,2]
        sort(nums.begin(),nums.end());
        for(int i=0;i<n;++i){
            if(i!=0 && nums[i]==nums[i-1]) continue;
            int j=i+1,k=n-1;
            while(j<k){
                int sum=nums[i]+nums[j]+nums[k];
                if(sum==0){
                    res.push_back({nums[i],nums[j],nums[k]});
                    ++j;
                    --k;
                    while(j<n && nums[j]==nums[j-1]) ++j;
                    while(k>0 && nums[k]==nums[k+1]) --k;
                }
                else if(sum<0){
                    ++j;
                }
                else {
                    --k;
                } 
            }
        }
        return res;
    }
};

// [1,0,1,2,-1,-4]


// sort->[-4,-1,-,1,1,2]
// i,j,k